### ISSUE TITLE
RCE upon joining modded worlds

### ISSUE THREAD

**Issue by xxxmeh** [bluemagic123/tModLoader#2872]
### Version

Latest 1.4-stable

### OS

Linux

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

Me and a few other players were playing terraria together after being asked to play calamity on discord. We joined their world and were instantly ratted affecting 4 players. Tmodloader by default, synced and ran the hosts malicious mod. I believe that syncing mods is useful but this should be limited to steam workshop mods and disabled by default. I redownloaded tmodloader after resetting my computer and noticed that auto-sync and downloading only signed mods are options that exist but have awful defaults. Please change the default of auto-syncing to off and the default of only downloading signed mods as enabled. As a player who just downloaded tmodloader to join some randoms to play for fun, I didn't see anything about automatically syncing malware until it was too late.

### Steps to reproduce

1. join world with default tmodloader settings
2. ratted.

### Expected Behavior

unsigned mods not auto-synced and run on my computer by default

### Actual Behavior

malware running on my computer

### Reproduction frequency

100% of the time

### Additional Information

please just change the defaults. most of us are dumb and will run in with defaults trusting that using steam defaults will keep us safe.

**Comment by G0dWeilder** [comment]
I agree this should be a default option I was also ratted with a group of friends this way.

**Comment by Chicken-Bones** [comment]
This definitely need some adjustment, but unfortunately the "only download signed mods" feature is useless since the move to steam workshop. We'll be looking into a "only download mods from steam workshop" option, but even that won't help you if someone publishes a malicious mod (though you could report it).

We could consider a warning and prompt to change a setting when a mod has less than 1k subscribers on steam workshop or something, but all of these rely on a connection to the workshop, which isn't always available. 

Similarly, a server may be running a different version of a mod to what is on steam workshop, and there would be no way to transmit that to you. If a mod updates and the server does not, immediately, then you would be unable to connect until the server was restarted.

**Comment by Chicken-Bones** [comment]
The only good option is some sort of signing service, where a mod author proves that they have ownership of a mod on the workshop, and can get new versions of their mod signed before upload. If someone posts a malicious mod, then their account gets banned from the workshop.

Unfortunately, that wouldn't prevent their already signed malicious mod from being distributed.

This could be combined with "only download workshop mods" to check if the mod is still active on the workshop (hasn't been banned).

**Comment by Chicken-Bones** [comment]
Thus, for a mod to be trusted:
1. It must have a signature which matches the contents, including verifying the steam workshop contentid of the mod. 
2. The signing service must verify ownership of the steam account linked to the workshop item being signed
3. The workshop item must still exist (though the version may be different)

I have no idea how do implement item 2 at this stage.

**Comment by Chicken-Bones** [comment]
https://www.youtube.com/watch?v=mgtvLaPYL-I hmmmm

**Comment by drunderscore** [comment]
I was personally forwarded this video by a concerned friend about two weeks ago... the title has even been changed.

I personally described this video as "scareware" -- all the bad details, with not even a hint of showing the option to disable downloading, or the likelyhood of it's occurrence.

(edit: I don't wish to cause problems with the videos creator, but saying _"In this video I'm explaining how it works and how to protect yourself against dangerous mods like that."_ and not showing any options is quite misleading)

I do agree the default may want to be changed to disabled, or perhaps require a prompt when connecting to accept any mods.

**Comment by xxxmeh** [comment]
Video is exactly what happened to me and friends :(

Having the auto-sync disabled by default and a warning upon enabling it would likely prevent this from happening to a fair amount of people. I think preventing join over possibility of malware is better than just downloading and running unknown mods. As long as a list of missing mods is provided, workshop makes it fairly easy for the user to just download them on their own, which is a slightly evil opinion, but any prevention is better than joining someones world and unknowingly ending up with malware on your pc.

**Comment by Chicken-Bones** [comment]
We will definitely display a warning before downloading all mods from the server, along with a mods list, and the option to subscribe via steam instead. 

Note however, that the server may have old versions of the mods that aren't on steam and it needs to send you, and that people may put a malicious mod in place of an actual mod to pretend that it's just an old version.

**Comment by Chicken-Bones** [comment]
I'll be working on this next week.

**Comment by steviegt6** [comment]
> [youtube.com/watch?v=mgtvLaPYL-I](https://www.youtube.com/watch?v=mgtvLaPYL-I) hmmmm

I really love how I've only heard of these occurrences after this video came out; wonderful.

@Chicken-Bones, I believe the best thing to do is allow users to review which files are going to be downloaded to their computers, as well as an option to out-right disable syncing. Due to the nature of tModLoader, pretty much any measures aside from just telling the user what is attempted to be downloaded will do very little. The main advantage of multiplayer syncing is ensuring version parity between clients and the server, so downloading from the workshop is unreliable. Like always, the best solution is warning users to exercise caution when dealing with software that can run anything on your machine...

**Comment by G0dWeilder** [comment]
Yeah really hope this gets fixed just some sort of warning before joining a server / player would be incredible that just says something like "Hey you are not currently subscribed to x mods would you like to? If not you will be unable to play this world" or something along the lines of that if you said yes it would lead you to there workshop pages and if they dont have one it would just give warning abt it being unsigned or something like that idk just dont want this happening to anyone else just mainly posting this reply so this doesnt get forgotten about.

**Comment by acidvegas** [comment]
lol @ an rce being ignored for 3 years now.

**Comment by steviegt6** [comment]
The RCE is "you can run mods", that is what a mod loader does.

**Comment by acidvegas** [comment]
> The RCE is "you can run mods", that is what a mod loader does.

nice brain.

**Comment by steviegt6** [comment]
PR support for sandboxing .NET assemblies if you'd like.

**Comment by Chicken-Bones** [comment]
This has been mitigated as far as we are able via #4801 and #4345
